### ISSUE TITLE
XLA: Results mismatch between different tilings

### ISSUE THREAD

**Issue by rrozb** [sanchit-gandhi/whisper-jax#97]
Hi, 
When I am running following code:
```python
from whisper_jax import FlaxWhisperPipline
import jax.numpy as jnp
pipeline = FlaxWhisperPipline("openai/whisper-large-v2", dtype=jnp.bfloat16, batch_size=1)
text = pipeline("audio.mp3", task="transcribe", return_timestamps=True)
text = pipeline("audio.mp3")
```

I am getting the following message:

```python
2023-05-23 12:15:49.152818: E external/xla/xla/service/gpu/buffer_comparator.cc:731] Difference at 45: 348 vs 388
2023-05-23 12:15:49.152841: E external/xla/xla/service/gpu/buffer_comparator.cc:731] Difference at 299: 352 vs 394
2023-05-23 12:15:49.152846: E external/xla/xla/service/gpu/buffer_comparator.cc:731] Difference at 407: 346 vs 394
2023-05-23 12:15:49.152850: E external/xla/xla/service/gpu/buffer_comparator.cc:731] Difference at 460: 354 vs 394
2023-05-23 12:15:49.152853: E external/xla/xla/service/gpu/buffer_comparator.cc:731] Difference at 527: 350 vs 390
2023-05-23 12:15:49.152856: E external/xla/xla/service/gpu/buffer_comparator.cc:731] Difference at 533: 352 vs 396
2023-05-23 12:15:49.152859: E external/xla/xla/service/gpu/buffer_comparator.cc:731] Difference at 553: 348 vs 392
2023-05-23 12:15:49.152863: E external/xla/xla/service/gpu/buffer_comparator.cc:731] Difference at 593: 348 vs 394
2023-05-23 12:15:49.152867: E external/xla/xla/service/gpu/buffer_comparator.cc:731] Difference at 694: 344 vs 390
2023-05-23 12:15:49.152870: E external/xla/xla/service/gpu/buffer_comparator.cc:731] Difference at 721: 354 vs 396
2023-05-23 12:15:49.152874: E external/xla/xla/service/gpu/triton_autotuner.cc:377] Results mismatch between different tilings. This is likely a bug/unexpected loss of precision.
2023-05-23 12:15:49.157692: E external/xla/xla/service/gpu/buffer_comparator.cc:731] Difference at 45: 348 vs 388
2023-05-23 12:15:49.157704: E external/xla/xla/service/gpu/buffer_comparator.cc:731] Difference at 299: 352 vs 394
2023-05-23 12:15:49.157708: E external/xla/xla/service/gpu/buffer_comparator.cc:731] Difference at 407: 346 vs 394
2023-05-23 12:15:49.157712: E external/xla/xla/service/gpu/buffer_comparator.cc:731] Difference at 460: 354 vs 394
2023-05-23 12:15:49.157716: E external/xla/xla/service/gpu/buffer_comparator.cc:731] Difference at 527: 350 vs 390
2023-05-23 12:15:49.157719: E external/xla/xla/service/gpu/buffer_comparator.cc:731] Difference at 533: 352 vs 396
2023-05-23 12:15:49.157722: E external/xla/xla/service/gpu/buffer_comparator.cc:731] Difference at 553: 348 vs 392
2023-05-23 12:15:49.157726: E external/xla/xla/service/gpu/buffer_comparator.cc:731] Difference at 593: 348 vs 394
2023-05-23 12:15:49.157729: E external/xla/xla/service/gpu/buffer_comparator.cc:731] Difference at 694: 344 vs 390
2023-05-23 12:15:49.157733: E external/xla/xla/service/gpu/buffer_comparator.cc:731] Difference at 721: 354 vs 396
2023-05-23 12:15:49.157737: E external/xla/xla/service/gpu/triton_autotuner.cc:377] Results mismatch between different tilings. This is likely a bug/unexpected loss of precision.
```

This however doesn't break code execution. I am able to get a transcription. The problem is that I don't see any increase in performance (slight decrease actually) compared with OpenAI implementation.  
10 min of audio takes 160 seconds to transcribe with JAX versus 115 seconds with OpenAI.
Could that performance issue be because of the error message? Does anyone know how to fix that error message?

I am running it with:

* Ubuntu version: 20.4
* JAX verison: 0.4.10
* Python version: 3.9.16
* GPU: NVIDIA GeForce RTX 3060
* Driver Version: 530.30.02
* CUDA Version: 12.1

**Comment by AvivNavon** [comment]
Hi, I'm experiencing this issue.

**Comment by Ordinath** [comment]
HI there! same for me. The weird thing is that it doesn't block the execution, so I tend to ignore this.

```
2023-05-29 13:29:32.586522  -  init pipeline
2023-05-29 13:29:52.106943: E external/xla/xla/service/gpu/buffer_comparator.cc:731] Difference at 45: 346.347 vs 387.222
2023-05-29 13:29:52.106981: E external/xla/xla/service/gpu/buffer_comparator.cc:731] Difference at 70: 350.232 vs 389.543
2023-05-29 13:29:52.106985: E external/xla/xla/service/gpu/buffer_comparator.cc:731] Difference at 299: 350.822 vs 394.675
2023-05-29 13:29:52.106988: E external/xla/xla/service/gpu/buffer_comparator.cc:731] Difference at 407: 345.434 vs 394.303
2023-05-29 13:29:52.107001: E external/xla/xla/service/gpu/buffer_comparator.cc:731] Difference at 533: 352.764 vs 396.02
2023-05-29 13:29:52.107003: E external/xla/xla/service/gpu/buffer_comparator.cc:731] Difference at 553: 347.322 vs 391.286
2023-05-29 13:29:52.107006: E external/xla/xla/service/gpu/buffer_comparator.cc:731] Difference at 593: 348.782 vs 393.536
2023-05-29 13:29:52.107009: E external/xla/xla/service/gpu/buffer_comparator.cc:731] Difference at 694: 343.945 vs 390.099
2023-05-29 13:29:52.107011: E external/xla/xla/service/gpu/buffer_comparator.cc:731] Difference at 721: 355.291 vs 395.885
2023-05-29 13:29:52.107014: E external/xla/xla/service/gpu/buffer_comparator.cc:731] Difference at 797: 347.525 vs 390.146
2023-05-29 13:29:52.107018: E external/xla/xla/service/gpu/triton_autotuner.cc:377] Results mismatch between different tilings. This is likely a bug/unexpected loss of precision.
2023-05-29 13:29:52.126014: E external/xla/xla/service/gpu/buffer_comparator.cc:731] Difference at 45: 346.347 vs 387.222
2023-05-29 13:29:52.126052: E external/xla/xla/service/gpu/buffer_comparator.cc:731] Difference at 70: 350.232 vs 389.543
2023-05-29 13:29:52.126056: E external/xla/xla/service/gpu/buffer_comparator.cc:731] Difference at 299: 350.822 vs 394.675
2023-05-29 13:29:52.126059: E external/xla/xla/service/gpu/buffer_comparator.cc:731] Difference at 407: 345.434 vs 394.303
2023-05-29 13:29:52.126071: E external/xla/xla/service/gpu/buffer_comparator.cc:731] Difference at 533: 352.764 vs 396.02
2023-05-29 13:29:52.126075: E external/xla/xla/service/gpu/buffer_comparator.cc:731] Difference at 553: 347.322 vs 391.286
2023-05-29 13:29:52.126077: E external/xla/xla/service/gpu/buffer_comparator.cc:731] Difference at 593: 348.782 vs 393.536
2023-05-29 13:29:52.126081: E external/xla/xla/service/gpu/buffer_comparator.cc:731] Difference at 694: 343.945 vs 390.099
2023-05-29 13:29:52.126084: E external/xla/xla/service/gpu/buffer_comparator.cc:731] Difference at 721: 355.291 vs 395.885
2023-05-29 13:29:52.126093: E external/xla/xla/service/gpu/buffer_comparator.cc:731] Difference at 797: 347.525 vs 390.146
2023-05-29 13:29:52.126100: E external/xla/xla/service/gpu/triton_autotuner.cc:377] Results mismatch between different tilings. This is likely a bug/unexpected loss of precision.
```

- Ubuntu version: 20.4
- JAX verison: 0.4.10
- Python version: 3.10.11
- GPU: NVIDIA GeForce RTX 4090
- NVIDIA-SMI 530.51.01
- Driver Version: 532.03
- CUDA Version: 12.1

**Comment by jelmers19** [comment]
Hi,

for me exactly the same. when i run from the python interpreter, i get the warnings below. however, when i run from a jupyter notebook, the kernel dies each time and i can't run my script

```
2023-05-30 09:25:19.747302: E external/xla/xla/service/gpu/buffer_comparator.cc:731] Difference at 1: 36.0938 vs 40.2812
2023-05-30 09:25:19.747332: E external/xla/xla/service/gpu/buffer_comparator.cc:731] Difference at 36: 35.0625 vs 39.7812
2023-05-30 09:25:19.747342: E external/xla/xla/service/gpu/buffer_comparator.cc:731] Difference at 90: 35.625 vs 39.7188
2023-05-30 09:25:19.747350: E external/xla/xla/service/gpu/buffer_comparator.cc:731] Difference at 138: 35.4375 vs 39.7188
2023-05-30 09:25:19.747358: E external/xla/xla/service/gpu/buffer_comparator.cc:731] Difference at 199: 35.0938 vs 39.5312
2023-05-30 09:25:19.747366: E external/xla/xla/service/gpu/buffer_comparator.cc:731] Difference at 250: 35.4375 vs 39.5312
2023-05-30 09:25:19.747371: E external/xla/xla/service/gpu/buffer_comparator.cc:731] Difference at 251: 35.625 vs 39.875
2023-05-30 09:25:19.747377: E external/xla/xla/service/gpu/buffer_comparator.cc:731] Difference at 281: 34.6562 vs 39.3438
2023-05-30 09:25:19.747383: E external/xla/xla/service/gpu/buffer_comparator.cc:731] Difference at 284: 35.8438 vs 40.0625
2023-05-30 09:25:19.747388: E external/xla/xla/service/gpu/buffer_comparator.cc:731] Difference at 317: 35.5 vs 40.0625
2023-05-30 09:25:19.747395: E external/xla/xla/service/gpu/triton_autotuner.cc:377] Results mismatch between different tilings. This is likely a bug/unexpected loss of precision.
2023-05-30 09:25:19.751579: E external/xla/xla/service/gpu/buffer_comparator.cc:731] Difference at 1: 36.0938 vs 40.2812
2023-05-30 09:25:19.751598: E external/xla/xla/service/gpu/buffer_comparator.cc:731] Difference at 36: 35.0625 vs 39.7812
2023-05-30 09:25:19.751606: E external/xla/xla/service/gpu/buffer_comparator.cc:731] Difference at 90: 35.625 vs 39.7188
2023-05-30 09:25:19.751612: E external/xla/xla/service/gpu/buffer_comparator.cc:731] Difference at 138: 35.4375 vs 39.7188
2023-05-30 09:25:19.751618: E external/xla/xla/service/gpu/buffer_comparator.cc:731] Difference at 199: 35.0938 vs 39.5312
2023-05-30 09:25:19.751626: E external/xla/xla/service/gpu/buffer_comparator.cc:731] Difference at 250: 35.4375 vs 39.5312
2023-05-30 09:25:19.751631: E external/xla/xla/service/gpu/buffer_comparator.cc:731] Difference at 251: 35.625 vs 39.875
2023-05-30 09:25:19.751637: E external/xla/xla/service/gpu/buffer_comparator.cc:731] Difference at 281: 34.6562 vs 39.3438
2023-05-30 09:25:19.751643: E external/xla/xla/service/gpu/buffer_comparator.cc:731] Difference at 284: 35.8438 vs 40.0625
2023-05-30 09:25:19.751649: E external/xla/xla/service/gpu/buffer_comparator.cc:731] Difference at 317: 35.5 vs 40.0625
2023-05-30 09:25:19.751655: E external/xla/xla/service/gpu/triton_autotuner.cc:377] Results mismatch between different tilings. This is likely a bug/unexpected loss of precision.
```

- Ubuntu version 20.04
- JAX version: 0.4.10
- Python version: 3.9.16
- GPU: NVIDIA TU104GL (Tesla T4)
- NVIDIA-SMI 470.182.03
- Driver Version: 470.182.03
- CUDA Version: 11.4

**Comment by lukeewin** [comment]
I also encountered this problem, did you finally solve it?

`2023-05-31 20:26:49.675032: E external/xla/xla/service/gpu/buffer_comparator.cc:731] Difference at 45: 346.461 vs 387.346
2023-05-31 20:26:49.675074: E external/xla/xla/service/gpu/buffer_comparator.cc:731] Difference at 70: 350.349 vs 389.669
2023-05-31 20:26:49.675084: E external/xla/xla/service/gpu/buffer_comparator.cc:731] Difference at 299: 350.939 vs 394.799
2023-05-31 20:26:49.675092: E external/xla/xla/service/gpu/buffer_comparator.cc:731] Difference at 407: 345.55 vs 394.43
2023-05-31 20:26:49.675101: E external/xla/xla/service/gpu/buffer_comparator.cc:731] Difference at 533: 352.882 vs 396.151
2023-05-31 20:26:49.675108: E external/xla/xla/service/gpu/buffer_comparator.cc:731] Difference at 553: 347.435 vs 391.412
2023-05-31 20:26:49.675116: E external/xla/xla/service/gpu/buffer_comparator.cc:731] Difference at 593: 348.895 vs 393.663
2023-05-31 20:26:49.675124: E external/xla/xla/service/gpu/buffer_comparator.cc:731] Difference at 694: 344.056 vs 390.228
2023-05-31 20:26:49.675132: E external/xla/xla/service/gpu/buffer_comparator.cc:731] Difference at 721: 355.399 vs 396.015
2023-05-31 20:26:49.675140: E external/xla/xla/service/gpu/buffer_comparator.cc:731] Difference at 797: 347.641 vs 390.267
2023-05-31 20:26:49.675148: E external/xla/xla/service/gpu/triton_autotuner.cc:377] Results mismatch between different tilings. This is likely a bug/unexpected loss of precision.
2023-05-31 20:26:49.706819: E external/xla/xla/service/gpu/buffer_comparator.cc:731] Difference at 45: 346.461 vs 387.346
2023-05-31 20:26:49.706848: E external/xla/xla/service/gpu/buffer_comparator.cc:731] Difference at 70: 350.349 vs 389.669
2023-05-31 20:26:49.706858: E external/xla/xla/service/gpu/buffer_comparator.cc:731] Difference at 299: 350.939 vs 394.799
2023-05-31 20:26:49.706866: E external/xla/xla/service/gpu/buffer_comparator.cc:731] Difference at 407: 345.55 vs 394.43
2023-05-31 20:26:49.706874: E external/xla/xla/service/gpu/buffer_comparator.cc:731] Difference at 533: 352.882 vs 396.151
2023-05-31 20:26:49.706882: E external/xla/xla/service/gpu/buffer_comparator.cc:731] Difference at 553: 347.435 vs 391.412
2023-05-31 20:26:49.706890: E external/xla/xla/service/gpu/buffer_comparator.cc:731] Difference at 593: 348.895 vs 393.663
2023-05-31 20:26:49.706898: E external/xla/xla/service/gpu/buffer_comparator.cc:731] Difference at 694: 344.056 vs 390.228
2023-05-31 20:26:49.706906: E external/xla/xla/service/gpu/buffer_comparator.cc:731] Difference at 721: 355.399 vs 396.015
2023-05-31 20:26:49.706914: E external/xla/xla/service/gpu/buffer_comparator.cc:731] Difference at 797: 347.641 vs 390.267
2023-05-31 20:26:49.706923: E external/xla/xla/service/gpu/triton_autotuner.cc:377] Results mismatch between different tilings. This is likely a bug/unexpected loss of precision.`

* CentOS 7.9
* NVIDIA-SMI: 530.41.03
* Driver Version: 530.41.03

**Comment by RichFree** [comment]
To anyone who comes across this from a google search, try increasing the matmul precision to at least "high"
```python
jax.config.update("jax_default_matmul_precision", "high")  
```

**Comment by danielracz** [comment]
> To anyone who comes across this from a google search, try increasing the matmul precision to at least "high"
> 
> ```python
> jax.config.update("jax_default_matmul_precision", "high")  
> ```

Thank  you, this seems to work for me so far! How did you find this solution?

**Comment by RichFree** [comment]
jax seems to default to a matmul precision of "low", and seeing how the error is related to a loss of precision, it seems reasonable to raise the matmul precision to "high" to alleviate the issue.

But after updating both jax and the nvidia drivers to the most up-to-date versions, setting the matmul precision to "low" works again. I suspect it could be a bug relating to bf16 matmul in the earlier versions, though I'm not sure if its jax or the gpu driver that was at fault.
